File: go-patroni/internal/fileperm/fileperm.go
// Package fileperm provides helpers for managing PostgreSQL file and directory permissions.
package fileperm

import (
	"os"
	"sync"

	"github.com/rs/zerolog/log"
)

const (
	// Mode mask for data directory permissions that only allows the owner - mask 077
	pgModeMaskOwner = 0o077

	// Mode mask for data directory permissions that allows group read/execute - mask 027
	pgModeMaskGroup = 0o027

	// Default mode for creating directories - mode 700
	pgDirModeOwner = 0o700

	// Mode for creating directories that allows group read/execute - mode 750
	pgDirModeGroup = 0o750

	// Default mode for creating files - mode 600
	pgFileModeOwner = 0o600

	// Mode for creating files that allows group read - mode 640
	pgFileModeGroup = 0o640
)

// FilePermissions manages permissions for directories and files under PGDATA.
type FilePermissions struct {
	mu               sync.RWMutex
	dirCreateMode    os.FileMode
	fileCreateMode   os.FileMode
	modeMask         int
	origUmask        int
}

// New creates a new FilePermissions with default owner-only permissions.
func New() *FilePermissions {
	fp := &FilePermissions{}
	fp.setOwnerPermissions()
	fp.origUmask = fp.setUmask()
	return fp
}

// setUmask sets the umask value based on current mode mask.
func (fp *FilePermissions) setUmask() int {
	oldUmask := setUmask(fp.modeMask)
	return oldUmask
}

// setOwnerPermissions makes directories/files accessible only by the owner.
func (fp *FilePermissions) setOwnerPermissions() {
	fp.dirCreateMode = pgDirModeOwner
	fp.fileCreateMode = pgFileModeOwner
	fp.modeMask = pgModeMaskOwner
}

// setGroupPermissions makes directories/files accessible by owner and readable by group.
func (fp *FilePermissions) setGroupPermissions() {
	fp.dirCreateMode = pgDirModeGroup
	fp.fileCreateMode = pgFileModeGroup
	fp.modeMask = pgModeMaskGroup
}

// SetPermissionsFromDataDirectory sets permissions based on the PGDATA directory.
func (fp *FilePermissions) SetPermissionsFromDataDirectory(dataDir string) {
	fp.mu.Lock()
	defer fp.mu.Unlock()

	info, err := os.Stat(dataDir)
	if err != nil {
		log.Error().Err(err).Str("data_dir", dataDir).Msg("Cannot check permissions")
		return
	}

	mode := info.Mode().Perm()
	if (mode & pgDirModeGroup) == pgDirModeGroup {
		fp.setGroupPermissions()
	} else {
		fp.setOwnerPermissions()
	}

	fp.setUmask()
}

// DirCreateMode returns the mode to use when creating directories.
func (fp *FilePermissions) DirCreateMode() os.FileMode {
	fp.mu.RLock()
	defer fp.mu.RUnlock()
	return fp.dirCreateMode
}

// FileCreateMode returns the mode to use when creating files.
func (fp *FilePermissions) FileCreateMode() os.FileMode {
	fp.mu.RLock()
	defer fp.mu.RUnlock()
	return fp.fileCreateMode
}

// OrigUmask returns the original umask value before modification.
func (fp *FilePermissions) OrigUmask() int {
	fp.mu.RLock()
	defer fp.mu.RUnlock()
	return fp.origUmask
}

// IsSecure checks if a file mode is secure (no world access).
func IsSecure(mode os.FileMode) bool {
	// PostgreSQL data directory should not have world access
	return (mode.Perm() & 0o007) == 0
}

// setUmask is a platform-specific function to set umask.
// On Unix systems, this calls syscall.Umask.
func setUmask(mask int) int {
	// Note: On Windows, umask is not meaningful, return 0o022 as default
	return doSetUmask(mask)
}

// Default is the global default FilePermissions instance.
var Default = New()

File: patroni/file_perm.nim
## Helper object that helps with figuring out file and directory permissions based on permissions of PGDATA.
##
## :var pg_perm: instance of the FilePermissions object.

import std/[posix, strformat]
import ./log

let logger = getLogger("patroni.file_perm")

const
  # Mode mask for data directory permissions that only allows the owner to
  # read/write directories and files -- mask 077.
  PG_MODE_MASK_OWNER = Mode(S_IRWXG or S_IRWXO)

  # Mode mask for data directory permissions that also allows group read/execute -- mask 027.
  PG_MODE_MASK_GROUP = Mode(S_IWGRP or S_IRWXO)

  # Default mode for creating directories -- mode 700.
  PG_DIR_MODE_OWNER = Mode(S_IRWXU)

  # Mode for creating directories that allows group read/execute -- mode 750.
  PG_DIR_MODE_GROUP = Mode(S_IRWXU or S_IRGRP or S_IXGRP)

  # Default mode for creating files -- mode 600.
  PG_FILE_MODE_OWNER = Mode(S_IRUSR or S_IWUSR)

  # Mode for creating files that allows group read -- mode 640.
  PG_FILE_MODE_GROUP = Mode(S_IRUSR or S_IWUSR or S_IRGRP)

type
  FilePermissions* = ref object
    ## Helper class for managing permissions of directories and files under PGDATA.
    ##
    ## Execute setPermissionsFromDataDirectory to figure out which permissions should be used for files and
    ## directories under PGDATA based on permissions of PGDATA root directory.
    pgDirCreateMode: Mode
    pgFileCreateMode: Mode
    pgModeMask: Mode
    origUmask*: Mode

proc setOwnerPermissions(self: FilePermissions) =
  ## Make directories/files accessible only by the owner.
  self.pgDirCreateMode = PG_DIR_MODE_OWNER
  self.pgFileCreateMode = PG_FILE_MODE_OWNER
  self.pgModeMask = PG_MODE_MASK_OWNER

proc setGroupPermissions(self: FilePermissions) =
  ## Make directories/files accessible by the owner and readable by group.
  self.pgDirCreateMode = PG_DIR_MODE_GROUP
  self.pgFileCreateMode = PG_FILE_MODE_GROUP
  self.pgModeMask = PG_MODE_MASK_GROUP

proc setUmaskInternal(self: FilePermissions): Mode =
  ## Set umask value based on calculations.
  ##
  ## .. note::
  ##     Should only be called once either setOwnerPermissions
  ##     or setGroupPermissions has been executed.
  ##
  ## :returns: the previous value of the umask or 0o22 if umask call failed.
  try:
    result = umask(self.pgModeMask)
  except CatchableError as e:
    logger.error(fmt"Can not set umask to {self.pgModeMask:03o}: {e.msg}")
    result = Mode(0o22)

proc newFilePermissions*(): FilePermissions =
  ## Create a FilePermissions object and set default permissions.
  new(result)
  result.setOwnerPermissions()
  result.origUmask = result.setUmaskInternal()

proc setPermissionsFromDataDirectory*(self: FilePermissions, dataDir: string) =
  ## Set new permissions based on provided dataDir.
  ##
  ## :param dataDir: reference to PGDATA to calculate permissions from.
  var st: Stat
  try:
    if stat(dataDir.cstring, st) == 0:
      if (st.st_mode and PG_DIR_MODE_GROUP) == PG_DIR_MODE_GROUP:
        self.setGroupPermissions()
      else:
        self.setOwnerPermissions()
      discard self.setUmaskInternal()
    else:
      logger.error(fmt"Can not check permissions on {dataDir}: stat failed")
  except CatchableError as e:
    logger.error(fmt"Can not check permissions on {dataDir}: {e.msg}")

proc dirCreateMode*(self: FilePermissions): Mode =
  ## Directory permissions.
  result = self.pgDirCreateMode

proc fileCreateMode*(self: FilePermissions): Mode =
  ## File permissions.
  result = self.pgFileCreateMode

# Global instance
var pgPerm* = newFilePermissions()
